src/ir/auth_logic/ast_string_test.cc

```diff
@@ -0,0 +1,154 @@
+  EXPECT_EQ(ToString(relationDeclaration2),
+            ".decl attribute rel2(arg3 : someProgrammerType::CustomType)");
+}
+
+static Principal principal1("principal1");
+
+TEST(PrincipalToStringTest, SimplePrincipalTest) {
+  EXPECT_EQ(ToString(principal1), "principal1");
+}
+
+static Principal principal2("principal2");
+static Attribute attribute1(principal1, predicate1);
+
+TEST(AttributeToStringTest, SimpleAttributeTest) {
+  EXPECT_EQ(ToString(attribute1), "principal1 predicate1(arg1, arg2)");
+}
+
+static CanActAs canActAs1(principal1, principal2);
+
+TEST(CanActAsToStringTest, SimpleCanActAsTest) {
+  EXPECT_EQ(ToString(canActAs1), "principal1 canActAs principal2");
+}
+
+static BaseFact baseFact1(predicate1);
+static BaseFact baseFact2(attribute1);
+static BaseFact baseFact3(canActAs1);
+TEST(BaseFactToStringTest, SimpleBaseFactTest) {
+  EXPECT_EQ(ToString(baseFact1), "predicate1(arg1, arg2)");
+  EXPECT_EQ(ToString(baseFact2), "principal1 predicate1(arg1, arg2)");
+  EXPECT_EQ(ToString(baseFact3), "principal1 canActAs principal2");
+}
+
+static Fact fact1({}, baseFact1);
+static Fact fact2({principal1}, baseFact1);
+TEST(FactToStringTest, SimpleFactTest) {
+  EXPECT_EQ(ToString(fact1), "predicate1(arg1, arg2)");
+  EXPECT_EQ(ToString(fact2), "principal1 canSay predicate1(arg1, arg2)");
+}
+
+static ConditionalAssertion condAssertion1(fact1, {baseFact1});
+static ConditionalAssertion condAssertion2(fact1, {baseFact1, baseFact2});
+TEST(ConditionalAssertionToStringTest, SimpleCondAssertionTest) {
+  EXPECT_EQ(ToString(condAssertion1),
+            "predicate1(arg1, arg2) :- "
+            "predicate1(arg1, arg2)");
+  EXPECT_EQ(ToString(condAssertion2),
+            "predicate1(arg1, arg2) :- predicate1(arg1, "
+            "arg2), principal1 predicate1(arg1, arg2)");
+}
+
+static Assertion assertion1(fact1);
+static Assertion assertion2(condAssertion1);
+TEST(AssertionToStringTest, SimpleAssertionTest) {
+  EXPECT_EQ(ToString(assertion1), "predicate1(arg1, arg2).");
+  EXPECT_EQ(ToString(assertion2),
+            "predicate1(arg1, arg2) :- "
+            "predicate1(arg1, arg2).");
+}
+
+static SaysAssertion saysAssertion1(principal1, {assertion1});
+static SaysAssertion saysAssertion2(principal2, {assertion1, assertion2});
+TEST(SaysAssertionToStringTest, SimpleSaysAssertionTest) {
+  EXPECT_EQ(ToString(saysAssertion1),
+            "principal1 says {\npredicate1(arg1, arg2).}");
+  EXPECT_EQ(ToString(saysAssertion2),
+            "principal2 says {\npredicate1(arg1, "
+            "arg2).\npredicate1(arg1, arg2) :- "
+            "predicate1(arg1, arg2).}");
+}
+
+static Query query1("query1", principal1, fact1);
+static Query query2("query2", principal2, fact1);
+TEST(QueryToStringTest, SimpleQueryTest) {
+  EXPECT_EQ(ToString(query1),
+            "Query(query1 = principal1 says predicate1(arg1, "
+            "arg2)?)");
+  EXPECT_EQ(ToString(query2),
+            "Query(query2 = principal2 says predicate1(arg1, "
```

Comment:
Do we need `Query` in the output? If this is for human consumption, `?` at the end is sufficient to mark this as a query?

Reply:
fixed